homeassistant/components/timer.py

```diff
@@ -0,0 +1,324 @@
+
+        attr = None
+        if service.service == SERVICE_PAUSE:
+            attr = 'async_pause'
+        elif service.service == SERVICE_CANCEL:
+            attr = 'async_cancel'
+        elif service.service == SERVICE_FINISH:
+            attr = 'async_finish'
+
+        tasks = [getattr(timer, attr)() for timer in target_timers if attr]
+        if service.service == SERVICE_START:
+            for timer in target_timers:
+                tasks.append(
+                    timer.async_start(service.data.get(ATTR_DURATION))
+                )
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), os.path.join(DOMAIN, 'services.yaml'))
+    )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_START, async_handler_service,
+        descriptions[SERVICE_START], SERVICE_SCHEMA_DURATION)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PAUSE, async_handler_service,
+        descriptions[SERVICE_PAUSE], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_CANCEL, async_handler_service,
+        descriptions[SERVICE_CANCEL], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_FINISH, async_handler_service,
+        descriptions[SERVICE_FINISH], SERVICE_SCHEMA)
+
+    yield from component.async_add_entities(entities)
+    return True
+
+
+class Timer(Entity):
+    """Representation of a timer."""
+
+    def __init__(self, hass, object_id, name, icon, duration):
+        """Initialize a timer."""
+        self.entity_id = ENTITY_ID_FORMAT.format(object_id)
+        self._name = name
+        self._state = STATUS_IDLE
+        self._duration = duration
+        self._remaining = self._duration
+        self._icon = icon
+        self._hass = hass
+        self._end = None
+        self._listener = None
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return True
+
+    @property
+    def name(self):
+        """Return name of the timer."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the current value of the timer."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_DURATION: self._duration.__str__(),
+        }
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Call when entity is about to be added to Home Assistant."""
+        # If not None, we got an initial value.
+        if self._state is not None:
+            return
+
+        # pylint: disable=undefined-variable
+        state = yield from hass.helpers.restore_state.async_get_last_state(
```

Comment:
How can the tests pass if `hass` is not defined here and thus raises an exception? 🤔

Reply:
Because at runtime `hass` is available in the global scope. That's what I was confused about in your suggestion on how I should do it because that's not how it's done usually. I assume you meant I should go with `self._hass...` then. I'll do that later today.